### PR TITLE
Local Cartesian Option

### PR DESCRIPTION
**Pull request by AGummyBear**
This PR adds a feature to `navsat_transform` that allows you to optionally use a Local Cartesian (also known as a [Local Tangent Plane](https://en.wikipedia.org/wiki/Local_tangent_plane_coordinates)) coordinate system rather than the UTM coordinate system internally. This feature is useful for robots that operate within several kilometers of their origin location, but may need to operate close to or on top of a UTM boundary. Currently, `navsat_transform` does not handle crossing UTM boundaries.

The Local Cartesian transform is handled by GeographicLib: [Local Cartesian](https://geographiclib.sourceforge.io/html/classGeographicLib_1_1LocalCartesian.html)

Changes in this PR:
- Geographic lib is added as a build dependecy to the `robot_localization` package
- The parameter `use_local_cartesian` is added to `navsat_transform`
- Internal member variables of `navsat_transform` are re-named from "utm" to "cartesian" to more accurately reflect the data they hold

This feature has been tested on our robotic lawnmower across the UTM boundary between zones 16 and 17 at -84.0 degrees longitude.

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Is there a ticket that this is in response to?

**Comment by SteveMacenski** [comment]
I can't speak to the functionality, but from a code perspective, I have no issue with this.

**Comment by AGummyBear** [comment]
> Is there a ticket that this is in response to?

This is related to, but does not directly address #222

**Comment by SteveMacenski** [comment]
Can you clarify the differences? this is both for me and Tom, but also for posterity of people looking at this PR. A full description and important context is helpful.  

> Greenzie

Oh cool, I know of you guys. Hope all is well on that front.

**Comment by AGummyBear** [comment]
Sure. There has been some discussion already on ticket #222, but here's the gist of it:

The `navsat_transform` node within this package has the purpose of converting global GPS coordinates (in Latitude/Longitude) to a relative coordinate within the robot's world frame. This relative coordinate, and the robot's world frame, are both Cartesian coordinate systems. The UTM grid is a widely-known Cartesian representation of the Earth's surface, so it is useful for handling the conversion from Lat/Lon to Cartesian, and that is what `navsat_transform` currently uses internally. The problem is that because the Earth's surface is not flat, the UTM grid actually approximates the ellipsoid of the Earth using multiple flat zones, and each zone has its own coordinates and a unique origin. What this means is that coordinates in Zone 16 do not have a Cartesian relationship with coordinates in Zone 17. So if your robot were to cross a boundary between two UTM zones, `navsat_transform` is not able to handle the new conversion, since it's not a trivial relationship.

So UTM has issues that make it difficult to use `navsat_transform` near a zone boundary, but UTM is not the only Cartesian coordinate system for the Earth (it's just a very widely known and well-defined one). Because the Earth is so large, you can actually create a flat tangent-plane approximation centered at any point on the surface and use Cartesian math between coordinates. Unless you move a substantial distance (think tens to hundreds of km) from the origin of this Cartesian frame, your calculations remain accurate: moving 1m in the positive-Y direction is equivalent to moving 1m North. The change I've made for this PR allows you to select Local Cartesian as an alternative to UTM for internal coordinate conversions.

My change does not make `navsat_transform` robust to handling UTM boundaries, but it does provide a workaround that many people might find useful. Specifically,

- If your robot is operated at varied locations and may be near a UTM boundary
- If your robot will not travel over long distances from its origin location (the location at which it acquires a good GPS lock)
- If you do not use the option `broadcast_utm_transform` and specifically need it to be the UTM transform.
 
Then I recommend setting `use_local_cartesian=true`

For people whose robots typically operate in a few locations, and they know that these locations are not on a UTM boundary, it shouldn't make a difference whether they use Local Cartesian or UTM internally.

**Comment by SteveMacenski** [comment]
Awesome, makes sense, thanks for the overview. @ayrton04 should be around in the next few days (probably) to review but nothing sticks out to me as long as the transforms are correct.

**Comment by ayrton04** [comment]
Thank you very much for this PR. I will review it next week. In the meantime, the build farm is unhappy because of some linting:

http://build.ros.org/job/Mpr__robot_localization__ubuntu_bionic_amd64/37/testReport/roslint/LintCheck/roslint_robot_localization/

**Comment by AGummyBear** [comment]
Oops, I missed that one in my earlier fix. I ran the linter locally and it seems to be happy now.

File: include/robot_localization/navsat_transform.h
Comment (ayrton04) [inline include/robot_localization/navsat_transform.h]: With the method name change, the second and third parameters are now misaligned. Would you mind updating that?

File: package.xml
Comment (ayrton04) [inline package.xml]: I'd like to retain the alphabetical order of these, if possible.

File: src/navsat_transform.cpp
Comment (ayrton04) [inline src/navsat_transform.cpp]: Let's add two new parameters for `broadcast_cartesian_transform` and `broadcast_cartesian_transform_as_parent_frame` as well. Maybe check if the user set `broadcast_utm_transform` and issue a deprecation warning?

File: include/robot_localization/navsat_transform.h
Comment (ayrton04) [inline include/robot_localization/navsat_transform.h:53]: Nit: can we drop these headers before `<string>`? I like to keep the STL headers last.

File: src/navsat_transform.cpp
Comment (ayrton04) [inline src/navsat_transform.cpp]: The parentheses here are a little off. To retain the original style, we should do this:

`std::string cartesian_frame_id = (use_local_cartesian_? "local_enu" : "utm");`

I realize the parentheses are unnecessary, but IMO it makes these things more readable.

File: src/navsat_transform.cpp
Comment (ayrton04) [inline src/navsat_transform.cpp]: Same comment re: parameter alignment.

File: src/navsat_transform.cpp
Comment (ayrton04) [inline src/navsat_transform.cpp]: *robot_cartesian_pose

**Review by ayrton04 — APPROVED** [review]
Very cool! Approving pending the changes I've requested. Thanks for this, and sorry for the delay in reviewing.

File: include/robot_localization/navsat_transform.h
Comment (AGummyBear) [reply include/robot_localization/navsat_transform.h:53]: Done.

File: include/robot_localization/navsat_transform.h
Comment (AGummyBear) [reply include/robot_localization/navsat_transform.h]: Done.

File: package.xml
Comment (AGummyBear) [reply package.xml]: Done.

File: src/navsat_transform.cpp
Comment (AGummyBear) [reply src/navsat_transform.cpp]: Done.

File: src/navsat_transform.cpp
Comment (AGummyBear) [reply src/navsat_transform.cpp]: Done.

File: src/navsat_transform.cpp
Comment (AGummyBear) [reply src/navsat_transform.cpp]: I made this change and tested it locally, but I would like a second pair of eyes on it. Basically, if the old parameter is set, it issues a warning telling the user to use the new parameter. The value of the old parameters is not used, so they will revert to the defaults (`false`) for any user once they receive this update.

File: src/navsat_transform.cpp
Comment (ayrton04) [inline src/navsat_transform.cpp:99]: I think we should maybe warn them, but then set the equivalent `broadcast_cartesian_transform` setting anyway. It gets a little more complicated if they do something silly like set `broadcast_cartesian_transform: false` and then `broadcast_utm_transform: true` (or vice-versa), but there's no reason for this to get too complex.

Comment (AGummyBear) [reply src/navsat_transform.cpp:99]: Fair point. I've made the change to set the parameters when using the deprecated names.

Comment (ayrton04) [reply src/navsat_transform.cpp:99]: Sorry to be a pain, but I think this will overwrite the value of `broadcast_cartesian_transform_` with _false_ if they _don't_ have `broadcast_utm_transform` set. Maybe change this to `getParam`?

```
if (nh_priv.getParam("broadcast_utm_transform", broadcast_cartesian_transform_))
```

Comment (AGummyBear) [reply src/navsat_transform.cpp:99]: Yes, you are absolutely right. I changed that to `getParam()` and verified that it doesn't overwrite the values.

**Review by ayrton04 — APPROVED** [review]

**Comment by ayrton04** [comment]
Thanks again for the PR!

**Pull request merged**
